### ISSUE TITLE
Clarify the usage of Opencv and CameraBridge

### ISSUE THREAD

**Issue by GiulioRomualdi** [ami-iit/bipedal-locomotion-framework#531]
Today I was using `YarpCameraBridge` and I noticed one thing I would like to make clear for everyone uses blf.

Thanks to 

https://github.com/ami-iit/bipedal-locomotion-framework/blob/e8b99889abdb35887046e433cd19ec441ce4c988/src/RobotInterface/YarpImplementation/src/YarpCameraBridge.cpp#L544-L546 

and 

https://github.com/ami-iit/bipedal-locomotion-framework/blob/e8b99889abdb35887046e433cd19ec441ce4c988/src/RobotInterface/YarpImplementation/src/YarpCameraBridge.cpp#L595-L597

It is possible to retrieve the rgb and the depth images from a camera. The images are returned as `cv::Mat` object.
Now accordingly to opencv documentation `cv::Mat` is just

> a class with two data parts: the matrix header (containing information such as the size of the matrix, the method used for storing, at which address is the matrix stored, and so on) a pointer to the matrix containing the pixel values

This means that when one call `YarpCameraBridge::getColorImage()` the `colorImg` object is valid if and only if the memory pointed by `colorImg` is valid (`cv::Mat` is really similar to an `std::span<>`). 

Going deeper in the code, we can notice that given a rgb image, the following yarp command "converts" the yarp image into a opencv one.

https://github.com/ami-iit/bipedal-locomotion-framework/blob/e8b99889abdb35887046e433cd19ec441ce4c988/src/RobotInterface/YarpImplementation/src/YarpCameraBridge.cpp#L562

Looking at the [yarp implementation](https://www.yarp.it/git-master/Cv-inl_8h_source.html#l00086), we notice that
```cpp
template<typename T>
 ::cv::Mat yarp::cv::toCvMat(yarp::sig::ImageOf<T>& yarpImage)
 {
     ::cv::Mat outMat (yarpImage.height(), yarpImage.width(), yarp::cv::type_code<T>::value,
                       yarpImage.getRawImage(), yarpImage.getRowSize()); // RVO
     if (convert_code_to_cv<T>::value >= 0)
     {
         ::cv::cvtColor(outMat, outMat, convert_code_to_cv<T>::value);
     }
     return outMat;
 }
```
So a `outMat` is created with the following constructor `cv::Mat::Mat(int rows, int cols, int type, void* data, size_t step = AUTO_STEP)` as explained [here](https://docs.opencv.org/3.4/d3/d63/classcv_1_1Mat.html#a51615ebf17a64c968df0bf49b4de6a3a) 

> data: is a pointer to the user data. Matrix constructors that take data and step parameters do not allocate matrix data. Instead, they just initialize the matrix header that points to the specified data, which means that **no data is copied**. This operation is very efficient and can be used to process external data using OpenCV functions. The external data is not automatically deallocated, so you should take care of it.

So in the end the opencv image is valid only if `m_pimpl->rgbImage.first` is valid. This has two side effects:
1. It is not possible to retrieve all the images and the use it later this is valid for both `YarpCameraBridge::getColorImage()` and `YarpCameraBridge::getDepthImage`
2. `YarpCameraBridge` is **not** thread safe. 

Regarding the point 1 the following code should explain better the problem
```cpp
cv::Mat img1, img2;
cameraBridge.getColorImage("camera1", img1);
cameraBridge.getColorImage("camera2", img2);

// Now img1 is not animore valid
```

Regarding the second point, since both `getColorImage` and `getDepthImage` modifies rispectivelly 

https://github.com/ami-iit/bipedal-locomotion-framework/blob/e8b99889abdb35887046e433cd19ec441ce4c988/src/RobotInterface/YarpImplementation/src/YarpCameraBridge.cpp#L553

https://github.com/ami-iit/bipedal-locomotion-framework/blob/e8b99889abdb35887046e433cd19ec441ce4c988/src/RobotInterface/YarpImplementation/src/YarpCameraBridge.cpp#L567

https://github.com/ami-iit/bipedal-locomotion-framework/blob/e8b99889abdb35887046e433cd19ec441ce4c988/src/RobotInterface/YarpImplementation/src/YarpCameraBridge.cpp#L604

it is not possible to call from separed threads.

**Comment by GiulioRomualdi** [comment]
cc @ami-iit/blf-developers

**Comment by GiulioRomualdi** [comment]
This problem has been fixed by #516
